### PR TITLE
Collections refactoring

### PR DESCRIPTION
**Pull request by akallabeth**
@kubistika I think you´ll like the `ArrayList_ForEach` function ;)
a question though, the cleanups in proxy can also be done by setting a `ValueFree` function for the array list, any reason this is not used?

### CODE REVIEW

File: winpr/include/winpr/collections.h [hunk 1]
```diff
@@ -145,24 +115,24 @@ extern "C"
...
-	     index < ArrayList_Count(_lst) && (value = (_type)ArrayList_GetItem(_lst, index)); \
-	     index++)
+	WINPR_API BOOL ArrayList_ForEach(wArrayList* arrayList, ArrayList_ForEachFkt fkt, ...);
 
 	WINPR_API void ArrayList_Clear(wArrayList* arrayList);
-	WINPR_API BOOL ArrayList_Contains(wArrayList* arrayList, void* obj);
+	WINPR_API BOOL ArrayList_Contains(wArrayList* arrayList, const void* obj);
 
 	WINPR_API int ArrayList_Add(wArrayList* arrayList, void* obj);
 	WINPR_API BOOL ArrayList_Insert(wArrayList* arrayList, int index, void* obj);
```
Comment (kubistika) [inline winpr/include/winpr/collections.h:128]: Should it be changed to `const void* obj` like in the other fkts?

File: winpr/include/winpr/collections.h [hunk 2]
```diff
@@ -145,24 +115,24 @@ extern "C"
...
-	for (index = 0;                                                                        \
-	     index < ArrayList_Count(_lst) && (value = (_type)ArrayList_GetItem(_lst, index)); \
-	     index++)
+	WINPR_API BOOL ArrayList_ForEach(wArrayList* arrayList, ArrayList_ForEachFkt fkt, ...);
 
 	WINPR_API void ArrayList_Clear(wArrayList* arrayList);
-	WINPR_API BOOL ArrayList_Contains(wArrayList* arrayList, void* obj);
+	WINPR_API BOOL ArrayList_Contains(wArrayList* arrayList, const void* obj);
 
 	WINPR_API int ArrayList_Add(wArrayList* arrayList, void* obj);
```
Comment (kubistika) [inline winpr/include/winpr/collections.h:127]: Should it be changed to const void* obj like in the other fkts?

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by kubistika** [comment]
@freerdp-bot test

**Review by mfleisz — APPROVED** [review]
LGTM

**Pull request merged**